File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/text/composing/ComposingState.kt
package com.siansiansu.taigikeyboard.ime.text.composing

import com.siansiansu.taigikeyboard.ime.core.settings.ToneToggles
import com.siansiansu.taigikeyboard.ime.dictionary.TPSConverter
import com.siansiansu.taigikeyboard.ime.dictionary.ToneConverter
import com.siansiansu.taigikeyboard.ime.dictionary.ToneConverterModels

/**
 * Platform-neutral composing-buffer state machine.
 *
 * `rawInput` (original keystrokes, e.g. `"gua2"`) is the single source of
 * truth; [derivedDisplay] recomputes the tone-marked form via
 * [ToneConverter]. The wrapper layer ([ComposingManager] on Android) owns
 * `InputConnection` side effects and settings reads — this type stays
 * Kotlin-stdlib-pure so future shared-core extraction can reuse it verbatim.
 *
 * Every intent returns `(newState, ComposingTransition)`; the transition's
 * effects list is the platform-neutral side-effect script the wrapper must
 * execute in order. No hidden side channels.
 *
 * Kotlin mirrors iOS `Input/Composing/ComposingState.swift`. The Swift
 * version is a `mutating` struct; Kotlin idiom is an immutable `data class`
 * that returns a new state — semantically equivalent (see
 * `rules/android-guidelines.md` §1 Kotlin-to-Rust shape preferences).
 */
data class ComposingState(
    val phase: Phase = Phase.Idle,
    /**
     * `-1` in idle; `0` when a fresh composition begins; preserved across
     * [Intent.ReplaceLast] and candidate-bar taps. UI never writes this
     * directly — [apply] and [withSelectedIndex] are the only mutation
     * paths so the pure state cannot desync from the platform wrapper.
     */
    val selectedCandidateIndex: Int = -1,
) {
    sealed class Phase {
        object Idle : Phase()

        data class Composing(
            val raw: String,
        ) : Phase()
    }

    sealed class Intent {
        /** Begin a new composition buffer with [text] (caret reset). */
        data class Start(
            val text: String,
        ) : Intent()

        /** Append [char]. If currently idle, acts as [Start]. */
        data class Append(
            val char: String,
        ) : Intent()

        /** Convenience alias for `Append("-")` (TPS / romanization hyphenation). */
        object AppendHyphen : Intent()

        /**
         * Replace the last raw-input character (TPS auto-correct). Intentionally
         * preserves [selectedCandidateIndex] — unlike [Append], which resets
         * it to `0`.
         */
        data class ReplaceLast(
            val replacement: String,
        ) : Intent()

        /** Delete one grapheme. Exits to idle when the buffer empties. */
        object DeleteBackward : Intent()

        /** Commit the [derivedDisplay] form (tone-marked) to the document. */
        object CommitDerived : Intent()

        /**
         * Commit the literal `rawInput` (bypass tone conversion). Mirror of
         * iOS `.commitRaw`; no Android call-site triggers this today, but
         * the state supports it so pure-state tests match iOS.
         */
        object CommitRaw : Intent()

        /**
         * Commit [text] to the document, atomically replacing the current
         * preedit.
         */
        data class SelectSuggestion(
            val text: String,
        ) : Intent()

        /** Clear all state (e.g. keyboard teardown / mode switch). */
        object Reset : Intent()
    }

    val isComposing: Boolean
        get() = phase is Phase.Composing

    val rawInput: String
        get() = (phase as? Phase.Composing)?.raw ?: ""

    /**
     * Derive the display form for the current [rawInput]. TPS symbols are
     * already display-ready; POJ/TL route through [ToneConverter] with the
     * supplied toggles.
     *
     * Kept synchronous and pure so commit-path intents (see [apply]) can
     * attach the derived text to the emitted effect without extra
     * scheduling. Live-typing intents emit the raw preedit instead and
     * the wrapper re-derives off-main-thread (see
     * `composing-state-boundary.md` §11 Android addendum).
     */
    fun derivedDisplay(
        mode: ToneConverterModels.InputMode,
        toggles: ToneToggles,
    ): String {
        val raw = rawInput
        if (raw.isEmpty()) return ""
        if (TPSConverter.containsTPS(raw)) return raw
        return ToneConverter.convertToToneMarks(raw, mode, toggles)
    }

    /** Returns a new state with [selectedCandidateIndex] set to [index]. */
    fun withSelectedIndex(index: Int): ComposingState = copy(selectedCandidateIndex = index)

    /**
     * Apply an [intent] and emit the resulting transition.
     *
     * Returns `(newState, transition)`. The transition's [ComposingTransition.effects]
     * is the ordered side-effect script the wrapper executes against
     * [ComposingDelegate].
     *
     * Android-specific text-content note: [ComposingTransition.Effect.UpdatePreedit]
     * carries the RAW keystrokes for live-typing intents (Start / Append /
     * ReplaceLast / non-empty-result DeleteBackward). Commit-path intents
     * (CommitDerived / CommitRaw / SelectSuggestion) carry the synchronously
     * derived form in [ComposingTransition.Effect.CommitTextReplacingPreedit].
     * Rationale + iOS divergence documented in `composing-state-boundary.md`
     * §11 Android addendum.
     */
    fun apply(
        intent: Intent,
        mode: ToneConverterModels.InputMode,
        toggles: ToneToggles,
    ): Pair<ComposingState, ComposingTransition> =
        when (intent) {
            is Intent.Start -> {
                enterComposing(raw = intent.text, mode = mode, toggles = toggles)
            }

            is Intent.Append -> {
                if (phase is Phase.Idle) {
                    enterComposing(raw = intent.char, mode = mode, toggles = toggles)
                } else {
                    enterComposing(raw = rawInput + intent.char, mode = mode, toggles = toggles)
                }
            }

            Intent.AppendHyphen -> {
                apply(Intent.Append("-"), mode, toggles)
            }

            is Intent.ReplaceLast -> {
                if (phase !is Phase.Composing || rawInput.isEmpty()) {
                    noopTransition()
                } else {
                    // ReplaceLast preserves selectedCandidateIndex — it is a
                    // correction on top of an in-progress selection, not a
                    // fresh composition step (see iOS ComposingStateTests).
                    val newRaw = rawInput.dropLast(1) + intent.replacement
                    val newState = copy(phase = Phase.Composing(newRaw))
                    val transition =
                        ComposingTransition(
                            newPhase = newState.phase,
                            newSelectedIndex = selectedCandidateIndex,
                            effects =
                                listOf(
                                    ComposingTransition.Effect.UpdatePreedit(newRaw),
                                    ComposingTransition.Effect.PerformAutocomplete,
                                ),
                            // Live-typing: skip sync ToneConverter call on the
                            // main thread. Async derivation via
                            // CandidateUpdateCoordinator handles the derived
                            // form off Dispatchers.Default (see §11.10).
                            derivedDisplay = "",
                        )
                    newState to transition
                }
            }

            Intent.DeleteBackward -> {
                if (phase !is Phase.Composing || rawInput.isEmpty()) {
                    noopTransition()
                } else {
                    val newRaw = rawInput.dropLast(1)
                    if (newRaw.isEmpty()) {
                        exitToIdle(
                            effects =
                                listOf(
                                    ComposingTransition.Effect.ClearPreeditWithoutCommit,
                                    ComposingTransition.Effect.ResetAutocomplete,
                                    ComposingTransition.Effect.DeleteBackwardFromDocument,
                                ),
                        )
                    } else {
                        val newState = ComposingState(phase = Phase.Composing(newRaw), selectedCandidateIndex = 0)
                        val transition =
                            ComposingTransition(
                                newPhase = newState.phase,
                                newSelectedIndex = 0,
                                effects =
                                    listOf(
                                        ComposingTransition.Effect.UpdatePreedit(newRaw),
                                        ComposingTransition.Effect.PerformAutocomplete,
                                    ),
                                // Live-typing: skip sync derivation (§11.10).
                                derivedDisplay = "",
                            )
                        newState to transition
                    }
                }
            }

            Intent.CommitDerived -> {
                if (phase !is Phase.Composing) {
                    noopTransition()
                } else {
                    val derived = derivedDisplay(mode, toggles)
                    if (derived.isEmpty()) {
                        noopTransition()
                    } else {
                        exitToIdle(
                            effects =
                                listOf(
                                    ComposingTransition.Effect.CommitTextReplacingPreedit(derived),
                                    ComposingTransition.Effect.ResetAutocomplete,
                                    ComposingTransition.Effect.ResetAutocompleteContext,
                                ),
                        )
                    }
                }
            }

            Intent.CommitRaw -> {
                if (phase !is Phase.Composing || rawInput.isEmpty()) {
                    noopTransition()
                } else {
                    val text = rawInput
                    exitToIdle(
                        effects =
                            listOf(
                                ComposingTransition.Effect.CommitTextReplacingPreedit(text),
                                ComposingTransition.Effect.ResetAutocomplete,
                                ComposingTransition.Effect.ResetAutocompleteContext,
                            ),
                    )
                }
            }

            is Intent.SelectSuggestion -> {
                if (phase !is Phase.Composing) {
                    noopTransition()
                } else {
                    exitToIdle(
                        effects =
                            listOf(
                                ComposingTransition.Effect.CommitTextReplacingPreedit(intent.text),
                                ComposingTransition.Effect.ResetAutocomplete,
                                ComposingTransition.Effect.ResetAutocompleteContext,
                            ),
                    )
                }
            }

            Intent.Reset -> {
                if (phase is Phase.Idle) {
                    noopTransition()
                } else {
                    exitToIdle(
                        effects =
                            listOf(
                                ComposingTransition.Effect.ClearPreeditWithoutCommit,
                                ComposingTransition.Effect.ResetAutocomplete,
                            ),
                    )
                }
            }
        }

    private fun enterComposing(
        raw: String,
        @Suppress("UNUSED_PARAMETER") mode: ToneConverterModels.InputMode,
        @Suppress("UNUSED_PARAMETER") toggles: ToneToggles,
    ): Pair<ComposingState, ComposingTransition> {
        val newState = ComposingState(phase = Phase.Composing(raw), selectedCandidateIndex = 0)
        val transition =
            ComposingTransition(
                newPhase = newState.phase,
                newSelectedIndex = 0,
                effects =
                    listOf(
                        ComposingTransition.Effect.UpdatePreedit(raw),
                        ComposingTransition.Effect.PerformAutocomplete,
                    ),
                // Live-typing: skip sync ToneConverter call on the main
                // thread. The wrapper (ComposingManager) ignores
                // transition.derivedDisplay and the async derivation loop
                // (CandidateUpdateCoordinator.scheduleDisplayDerivation)
                // handles the tone-marked form on Dispatchers.Default. See
                // `composing-state-boundary.md` §11.10. `mode`/`toggles`
                // stay on the signature for call-site parity + future
                // shared-core alignment (iOS uses them here).
                derivedDisplay = "",
            )
        return newState to transition
    }

    private fun exitToIdle(effects: List<ComposingTransition.Effect>): Pair<ComposingState, ComposingTransition> {
        val newState = ComposingState(phase = Phase.Idle, selectedCandidateIndex = -1)
        val transition =
            ComposingTransition(
                newPhase = Phase.Idle,
                newSelectedIndex = -1,
                effects = effects,
                derivedDisplay = "",
            )
        return newState to transition
    }

    private fun noopTransition(): Pair<ComposingState, ComposingTransition> =
        this to
            ComposingTransition(
                newPhase = phase,
                newSelectedIndex = selectedCandidateIndex,
                effects = emptyList(),
                derivedDisplay = "",
            )
}

File: ios/Sources/TaigiKeyboard/Input/Composing/ComposingState.swift
import Foundation

// MARK: - Shared-Core Candidate

// Pure logic, Foundation-only. Eligible for cross-platform extraction.

/// Platform-neutral composing-buffer state machine.
///
/// `rawInput` (original keystrokes, e.g. `"gua2"`) is the single source of
/// truth; `derivedDisplay(...)` recomputes the tone-marked form via
/// `ToneConverter`. The wrapper layer (`ComposingManager` on iOS) owns
/// `@Published` fan-out, delegate side effects, and settings reads — this
/// type stays Foundation-pure so Android can reuse it verbatim.
///
/// Every intent returns a `ComposingTransition` whose `effects` the wrapper
/// must execute in the given order. No hidden side channels.
struct ComposingState: Equatable {
    enum Phase: Equatable {
        case idle
        case composing(raw: String)
    }

    enum Intent: Equatable {
        /// Begin a new composition buffer with the given text (caret reset).
        case start(String)

        /// Append `char` to the buffer; if idle, falls through to `start`.
        case append(String)

        /// Append `"-"` — convenience alias for TPS / romanization hyphenation.
        case appendHyphen

        /// Replace the last raw-input character (TPS auto-correct). Intentionally
        /// preserves `selectedCandidateIndex` — unlike `append`, which resets it.
        case replaceLast(String)

        /// Delete one grapheme. Exits to idle when the buffer empties.
        case deleteBackward

        /// Commit the `derivedDisplay` form (tone-marked) to the document.
        case commitDerived

        /// Commit the literal `rawInput` (bypass conversion). Used for
        /// Enter-at-index-0 / English passthrough.
        case commitRaw

        /// Commit the supplied suggestion text to the document, atomically
        /// replacing the current preedit.
        case selectSuggestion(String)

        /// Commit the current preedit (if any) and insert an externally
        /// supplied text atomically in one document write. Used by
        /// non-Taigi input surfaces — emoji palette, clipboard paste —
        /// so composing state never leaks a silent finish-composing.
        /// Idle → behaves as a plain insert of `text`.
        case commitPreeditThenInsertExternal(String)

        /// Clear all state (e.g. keyboard teardown / mode switch).
        case reset
    }

    private(set) var phase: Phase = .idle

    /// `-1` in idle; `0` when a fresh composition begins; preserved across
    /// `replaceLast`. UI never writes this directly — transitions are the
    /// only mutation path (see `composing-state-boundary.md` §2.1).
    private(set) var selectedCandidateIndex: Int = -1

    /// Externally override `selectedCandidateIndex` (e.g. candidate-bar tap
    /// or keyboard-arrow gesture). Kept as an explicit mutator rather than
    /// a writable property so all state changes flow through one of two
    /// paths — `apply(_:mode:toneToggles:)` for buffer intents, this method
    /// for UI-driven selection — and there is no way to desync the pure
    /// state from the platform wrapper.
    mutating func setSelectedCandidateIndex(_ index: Int) {
        selectedCandidateIndex = index
    }

    var isComposing: Bool {
        if case .idle = phase { return false }
        return true
    }

    var rawInput: String {
        if case let .composing(raw) = phase { return raw }
        return ""
    }

    /// Derive the display form for the current `rawInput`. TPS symbols are
    /// already display-ready; POJ/TL route through `ToneConverter` with the
    /// supplied toggles.
    func derivedDisplay(mode: InputMode, toneToggles: ToneToggles) -> String {
        let raw = rawInput
        guard !raw.isEmpty else { return "" }
        if RustEngineBridge.containsTPS(raw) { return raw }
        // Match the pre-D9.4 `ToneConverter.convertToToneMarks` pipeline:
        // normalize tones via the engine, then post-process the nasal
        // marker case so `ⁿ` / `ᴺ` follow the preceding letter's case
        // (uppercase before → `ᴺ`, otherwise `ⁿ`).
        let toneMarked = RustEngineBridge.normalizeTone(raw, mode: mode, toggles: toneToggles)
        return ToneUtilities.adjustNasalMarkerCase(toneMarked)
    }

    /// Apply an intent and emit the resulting `Transition`. The receiver is
    /// mutated in place; the returned transition describes the effects the
    /// wrapper must execute against the platform adapter.
    @discardableResult
    mutating func apply(
        _ intent: Intent,
        mode: InputMode,
        toneToggles: ToneToggles,
    ) -> ComposingTransition {
        switch intent {
        case let .start(text):
            return enterComposing(raw: text, mode: mode, toneToggles: toneToggles)

        case let .append(char):
            if case .idle = phase {
                return enterComposing(raw: char, mode: mode, toneToggles: toneToggles)
            }
            return enterComposing(raw: rawInput + char, mode: mode, toneToggles: toneToggles)

        case .appendHyphen:
            return apply(.append("-"), mode: mode, toneToggles: toneToggles)

        case let .replaceLast(replacement):
            guard case .composing = phase, !rawInput.isEmpty else {
                return noopTransition()
            }
            let newRaw = String(rawInput.dropLast()) + replacement
            // Replacement preserves selectedCandidateIndex — it is a correction
            // on top of an in-progress selection, not a fresh composition step.
            phase = .composing(raw: newRaw)
            let display = derivedDisplay(mode: mode, toneToggles: toneToggles)
            return ComposingTransition(
                newPhase: phase,
                newSelectedIndex: selectedCandidateIndex,
                effects: [.updatePreedit(display), .performAutocomplete],
                derivedDisplay: display,
            )

        case .deleteBackward:
            guard case .composing = phase, !rawInput.isEmpty else {
                return noopTransition()
            }
            let newRaw = String(rawInput.dropLast())
            if newRaw.isEmpty {
                return exitToIdle(
                    effects: [
                        .clearPreeditWithoutCommit,
                        .resetAutocomplete,
                        .deleteBackwardFromDocument,
                    ],
                )
            }
            // Keep engine state in sync with the transition's newSelectedIndex
            // (Codex P2 r3106434856): without this, a later `replaceLast` would
            // resurrect a pre-delete external selection because the engine's
            // stored value still holds the stale index.
            phase = .composing(raw: newRaw)
            selectedCandidateIndex = 0
            let display = derivedDisplay(mode: mode, toneToggles: toneToggles)
            return ComposingTransition(
                newPhase: phase,
                newSelectedIndex: 0,
                effects: [.updatePreedit(display), .performAutocomplete],
                derivedDisplay: display,
            )

        case .commitDerived:
            guard case .composing = phase else { return noopTransition() }
            let display = derivedDisplay(mode: mode, toneToggles: toneToggles)
            guard !display.isEmpty else { return noopTransition() }
            return exitToIdle(
                effects: [
                    .commitTextReplacingPreedit(display),
                    .resetAutocomplete,
                    .resetAutocompleteContext,
                ],
            )

        case .commitRaw:
            guard case .composing = phase, !rawInput.isEmpty else {
                return noopTransition()
            }
            let text = rawInput
            return exitToIdle(
                effects: [
                    .commitTextReplacingPreedit(text),
                    .resetAutocomplete,
                    .resetAutocompleteContext,
                ],
            )

        case let .selectSuggestion(text):
            guard case .composing = phase else { return noopTransition() }
            return exitToIdle(
                effects: [
                    .commitTextReplacingPreedit(text),
                    .resetAutocomplete,
                    .resetAutocompleteContext,
                ],
            )

        case let .commitPreeditThenInsertExternal(externalText):
            guard !externalText.isEmpty else { return noopTransition() }
            if case .composing = phase {
                let derived = derivedDisplay(mode: mode, toneToggles: toneToggles)
                return exitToIdle(
                    effects: [
                        .commitTextReplacingPreedit(derived + externalText),
                        .resetAutocomplete,
                        .resetAutocompleteContext,
                    ],
                )
            }
            // Idle: behave as a plain insert. Reuse `commitTextReplacingPreedit`
            // because it already maps to `clearMarkedText + insertText` on iOS
            // and `commitText` on Android; both are no-op-on-empty-preedit safe.
            return ComposingTransition(
                newPhase: .idle,
                newSelectedIndex: -1,
                effects: [.commitTextReplacingPreedit(externalText)],
                derivedDisplay: "",
            )

        case .reset:
            if case .idle = phase { return noopTransition() }
            return exitToIdle(effects: [.clearPreeditWithoutCommit, .resetAutocomplete])
        }
    }

    // MARK: - Private Helpers

    /// Enter (or update) the composing phase with the given raw buffer. A
    /// fresh composition step resets `selectedCandidateIndex` to `0`.
    private mutating func enterComposing(
        raw: String,
        mode: InputMode,
        toneToggles: ToneToggles,
    ) -> ComposingTransition {
        phase = .composing(raw: raw)
        selectedCandidateIndex = 0
        let display = derivedDisplay(mode: mode, toneToggles: toneToggles)
        return ComposingTransition(
            newPhase: phase,
            newSelectedIndex: 0,
            effects: [.updatePreedit(display), .performAutocomplete],
            derivedDisplay: display,
        )
    }

    /// Transition to idle, zeroing the selected index invariant, with the
    /// caller-specified effects (ordering depends on the intent).
    private mutating func exitToIdle(effects: [ComposingTransition.Effect]) -> ComposingTransition {
        phase = .idle
        selectedCandidateIndex = -1
        return ComposingTransition(
            newPhase: .idle,
            newSelectedIndex: -1,
            effects: effects,
            derivedDisplay: "",
        )
    }

    private func noopTransition() -> ComposingTransition {
        // Noop paths are reachable only from guard-idle branches, so the
        // derived display is empty by invariant. Wrapper keeps its last
        // published `composingText` untouched when `effects` is empty.
        ComposingTransition(
            newPhase: phase,
            newSelectedIndex: selectedCandidateIndex,
            effects: [],
            derivedDisplay: "",
        )
    }
}
